In [133]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

# Replace 'your_file.xlsx' with the actual file path
file_path = 'data/kpmg.xlsx'

# Use the ExcelFile class to read the Excel file
xls = pd.ExcelFile(file_path)

# List the sheet names in the Excel file
sheet_names = xls.sheet_names

# Create DataFrames for each sheet
dataframes = {}  # Dictionary to store DataFrames

for sheet_name in sheet_names:
    dataframes[sheet_name] = pd.read_excel(xls, sheet_name)

# Now you have separate DataFrames for each sheet
# access the dataframes using their sheet names
transaction_data = dataframes['Transactions']
demographic_data = dataframes['CustomerDemographic']
customer_data = dataframes['CustomerAddress']

In [134]:
# Reading the first five rows of the customer dataset
customer_data.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [135]:
# Reading the first five rows of the demographic dataset
demographic_data.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [136]:
# Get unique values in the 'gender' column and count the occurrences of each unique value
gender_counts = demographic_data['gender'].unique()
counts = demographic_data['gender'].value_counts()

# Print the unique values and their respective counts
print("Unique Gender Values:")
print(gender_counts)
print("\nGender Value Counts:")
print(counts)

Unique Gender Values:
['F' 'Male' 'Female' 'U' 'Femal' 'M']

Gender Value Counts:
Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: gender, dtype: int64


In [137]:
# Create a mapping dictionary to standardize gender values
gender_mapping = {
    'F': 'Female',
    'Male': 'Male',
    'Female': 'Female',
    'U': 'Unspecified',
    'Femal': 'Female',
    'M': 'Male'
}

# Use the mapping dictionary to replace gender values
demographic_data['gender'] = demographic_data['gender'].replace(gender_mapping)

# Check the corrected unique values and their counts
gender_counts_corrected = demographic_data['gender'].unique()
counts_corrected = demographic_data['gender'].value_counts()

# Print the corrected unique values and their counts
print("Unique Gender Values (Corrected):")
print(gender_counts_corrected)
print("\nGender Value Counts (Corrected):")
print(counts_corrected)

Unique Gender Values (Corrected):
['Female' 'Male' 'Unspecified']

Gender Value Counts (Corrected):
Female         2039
Male           1873
Unspecified      88
Name: gender, dtype: int64


In [138]:
# Reading the first five rows of the transaction dataset
transaction_data.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

## Accuracy and Completeness Assessment

***Check for data accuracy issues, such as missing values in critical columns.***

**customer Data**

In [139]:
customer_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [140]:
# Accuracy Assessment
import numpy as np
# Check for missing values in key columns
missing_values =customer_data.isnull().sum()

# Calculate the percentage of missing values in each column
total_records = customer_data.shape[0]
percentage_missing = (missing_values / total_records) * 100

# Print results
print(np.char.center('Missing Values in Transaction Dataset', 60, '*'))
print(missing_values)
print("\nPercentage of Missing Values:")
print(percentage_missing)

***********Missing Values in Transaction Dataset************
customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

Percentage of Missing Values:
customer_id           0.0
address               0.0
postcode              0.0
state                 0.0
country               0.0
property_valuation    0.0
dtype: float64


In [141]:
customer_data['property_valuation'].describe()

count    3999.000000
mean        7.514379
std         2.824663
min         1.000000
25%         6.000000
50%         8.000000
75%        10.000000
max        12.000000
Name: property_valuation, dtype: float64

In [142]:
duplicate_rows = customer_data[customer_data.duplicated()]
# Print duplicate rows, if any
print("Duplicate Rows in Transaction Dataset:")
print(duplicate_rows)


Duplicate Rows in Transaction Dataset:
Empty DataFrame
Columns: [customer_id, address, postcode, state, country, property_valuation]
Index: []


The `country` column in the 'customer_data' dataset contains the value "Australia" for all its entries. Since this column doesn't provide any additional information due to its uniformity (as it's the same for all records), we can safely remove it without losing any valuable data. This can help make your dataset more concise and easier to work with.

In [172]:
customer_data.drop('country', axis=1, inplace=True)

**Demographic_data**

In [143]:
demographic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [144]:
# Check for missing values in key columns
missing_values = demographic_data.isnull().sum()

# Calculate the percentage of missing values in each column
total_records = demographic_data.shape[0]
percentage_missing = (missing_values / total_records) * 100

# Print results
print(np.char.center('Missing Values in Demographic Dataset', 60, '*'))
print(missing_values)
print("\nPercentage of Missing Values:")
print(percentage_missing)

***********Missing Values in Demographic Dataset************
customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

Percentage of Missing Values:
customer_id                             0.000
first_name                              0.000
last_name                               3.125
gender                                  0.000
past_3_years_bike_related_purchases     0.000
DOB                                     2.175
job_title                              12.650
job_industry_c

`last_name:` This column has a relatively low percentage of missing values, about 3.1%. It's reasonable to fill the missing values with the mode, which means replacing them with the most frequently occurring last name in the dataset. This approach helps maintain data integrity.

`DOB (Date of Birth):` The 'DOB' column has around 2.2% missing values. Given the importance of this information, it's advisable to fill the missing values with the mode, which is the most common date of birth in the dataset. This way, you can preserve the completeness of the data.

`job_title:` The 'job_title' column has a higher percentage of missing values, approximately 12.6%. In this case, filling the missing values with the mode is a practical solution. It ensures that the most common job titles are used to complete the dataset, even though it may not be ideal for all cases.

`job_industry_category:` With about 16.4% missing values, the 'job_industry_category' column requires filling. Again, using the mode to replace missing values is a reasonable choice. This method ensures that the most prevalent industry categories are used for the missing entries.

`default:` The 'default' column has a relatively high percentage of missing values, around 7.6%. Filling these missing values with the mode is a suitable approach, making use of the most common default status. However, be aware that this method might not capture the complexity of individual default preferences.

`Tenure:` The 'tenure' column has approximately 2.2% missing values. Filling these missing values with the mean of the 'tenure' values is a practical solution. Using the mean allows you to maintain data continuity and helps provide an estimate for missing values based on the existing data.

In [145]:
# Fill missing values with mode for categorical columns
demographic_data['last_name'] = demographic_data['last_name'].fillna(demographic_data['last_name'].mode()[0])
demographic_data['DOB'] = demographic_data['DOB'].fillna(demographic_data['DOB'].mode()[0])
demographic_data['job_title'] = demographic_data['job_title'].fillna(demographic_data['job_title'].mode()[0])
demographic_data['job_industry_category'] = demographic_data['job_industry_category'].fillna(demographic_data['job_industry_category'].mode()[0])
demographic_data['default'] = demographic_data['default'].fillna(demographic_data['default'].mode()[0])

# Fill missing values in 'tenure' column with the mean
demographic_data['tenure'] = demographic_data['tenure'].fillna(demographic_data['tenure'].mean())

In [146]:
# Convert columns to the desired data types
demographic_data['past_3_years_bike_related_purchases'] = pd.to_numeric(demographic_data['past_3_years_bike_related_purchases'])
demographic_data['tenure'] = pd.to_numeric(demographic_data['tenure'])
demographic_data['DOB'] = pd.to_datetime(demographic_data['DOB'])

In [147]:
# Check for duplicates in the dataset
duplicates = demographic_data[demographic_data.duplicated()]

# Print the duplicated rows, if any
if not duplicates.empty:
    print("Duplicated Rows:")
    print(duplicates)
else:
    print("No duplicates found in the dataset.")

No duplicates found in the dataset.


In [148]:
# Collecting the categorical columns into a list
cat_col = demographic_data.select_dtypes(include=['object']).columns.tolist()
cat_col

['first_name',
 'last_name',
 'gender',
 'job_title',
 'job_industry_category',
 'wealth_segment',
 'deceased_indicator',
 'default',
 'owns_car']

In [149]:
for col in cat_col:
    print(f"{col}:")
    print(demographic_data[col].unique())
    print("*******")

first_name:
['Laraine' 'Eli' 'Arlin' ... 'Stephie' 'Rusty' 'Sarene']
*******
last_name:
['Medendorp' 'Bockman' 'Dearle' ... 'Halgarth' 'Woolley' 'Oldland']
*******
gender:
['Female' 'Male' 'Unspecified']
*******
job_title:
['Executive Secretary' 'Administrative Officer' 'Recruiting Manager'
 'Business Systems Development Analyst' 'Senior Editor' 'Media Manager I'
 'Senior Quality Engineer' 'Nuclear Power Engineer' 'Developer I'
 'Account Executive' 'Junior Executive' 'Media Manager IV'
 'Sales Associate' 'Professor' 'Geological Engineer' 'Project Manager'
 'Safety Technician I' 'Research Assistant I' 'Accounting Assistant III'
 'Editor' 'Research Nurse' 'Safety Technician III' 'Staff Accountant III'
 'Legal Assistant' 'Product Engineer' 'Information Systems Manager'
 'VP Quality Control' 'Social Worker' 'Senior Cost Accountant'
 'Assistant Media Planner' 'Payment Adjustment Coordinator' 'Food Chemist'
 'Accountant III' 'Director of Sales' 'Senior Financial Analyst'
 'Registered Nurse' 

In [150]:
demographic_data.drop('default', axis=1, inplace=True)


In [166]:
demographic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 12 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            4000 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  4000 non-null   datetime64[ns]
 6   job_title                            4000 non-null   object        
 7   job_industry_category                4000 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  owns_car    

To create a new 'Age' feature, we will convert the 'DOB' (Date of Birth) column into an `Age` column. Once this transformation is complete, we will drop the 'DOB' column from the dataset. 
This process involves replacing the date of birth values with the corresponding ages, and after successfully creating the new 'Age' feature, we remove the original 'DOB' column to keep our dataset more organized and concise

In [167]:
import datetime as dt

def from_dob_to_age(born):
    # Get today's date
    today = dt.date.today()
    
    # Calculate the age by subtracting the birth year
    age = today.year - born.year
    
    # Adjust the age if the birthdate hasn't occurred in the current year
    if (today.month, today.day) < (born.month, born.day):
        age -= 1
    
    return age

In [168]:
#applying the function on the DOB column
demographic_data['Age'] = demographic_data['DOB'].apply(lambda x: from_dob_to_age(x))

In [169]:
demographic_data.drop('DOB',axis=1,inplace=True)

In [170]:
# Print the information of our dataset
demographic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 12 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   int64  
 1   first_name                           4000 non-null   object 
 2   last_name                            4000 non-null   object 
 3   gender                               4000 non-null   object 
 4   past_3_years_bike_related_purchases  4000 non-null   int64  
 5   job_title                            4000 non-null   object 
 6   job_industry_category                4000 non-null   object 
 7   wealth_segment                       4000 non-null   object 
 8   deceased_indicator                   4000 non-null   object 
 9   owns_car                             4000 non-null   object 
 10  tenure                               4000 non-null   float64
 11  Age                           

In [171]:
# print the head of the dataset
demographic_data.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp  Female   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot   Pristnor    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases                             job_title  \
0                                   93                   Executive Secretary   
1                                   81                Administrative Officer   
2                                   61                    Recruiting Manager   
3                                   33  Business Systems Development Analyst   
4                                   56                         Senior Editor   

  job_industry_category     wealth_segment deceased_indicator owns_car  \
0                Health      Mass Customer                  N      Yes   
1    Financial Services      Mass Customer                  N      Yes   
2              Property      Mass Customer                  N      Yes   
3                    IT      Mass Customer                  N       No   
4         Manufacturing  Affluent Customer                  N      Yes   

   tenure  Age  
0    11.0   70  
1    16.0   42  
2    15.0   69  
3     7.0   62  
4     8.0   46

**Transaction Data**

In [151]:
transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [152]:
# Check for missing values in key columns
missing_values = transaction_data.isnull().sum()

# Calculate the percentage of missing values in each column
total_records = transaction_data.shape[0]
percentage_missing = (missing_values / total_records) * 100

# Print results
print(np.char.center('Missing Values in Demographic Dataset', 60, '*'))
print(missing_values)
print("\nPercentage of Missing Values:")
print(percentage_missing)

***********Missing Values in Demographic Dataset************
transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

Percentage of Missing Values:
transaction_id             0.000
product_id                 0.000
customer_id                0.000
transaction_date           0.000
online_order               1.800
order_status               0.000
brand                      0.985
product_line               0.985
product_class              0.985
product_size               0.985
list_price                 0.000
standard_cost              0.985
product_first_sold_date    0.985
dtype: float64


In [153]:
# filling the online order column with the mode value 
transaction_data['online_order'] = transaction_data['online_order'].fillna(transaction_data['online_order'].mode()[0])

`Online Order:` Since 'online_order' has only 1.8% missing values, it's a small proportion of the data. We can safely fill these missing values with the mode, which is the most frequent data point. Filling with the mode ensures that you don't lose important data while addressing the missing values.

`Brand`, `Product Line`, `Product Class`, `Product Size`, `Standard Cost`, and `Product First Sold Date`: These columns all have a 1% rate of missing values, and they seem to follow the same missing data pattern. In this case, it is reasonable to remove the rows containing these null values. The missing data accounts for only a small fraction of the dataset (1%), and removing them won't significantly impact your analysis or modeling process. This approach helps maintain data quality and consistency in the remaining dataset.

In [154]:
# Removing other data rows with nul values
transaction_data.dropna(axis=0,inplace=True)

In [155]:
# Check for missing values in key columns
missing_values = transaction_data.isnull().sum()

# Calculate the percentage of missing values in each column
total_records = transaction_data.shape[0]
percentage_missing = (missing_values / total_records) * 100

# Print results
print(np.char.center('Missing Values in Demographic Dataset', 60, '*'))
print(missing_values)
print("\nPercentage of Missing Values:")
print(percentage_missing)

***********Missing Values in Demographic Dataset************
transaction_id             0
product_id                 0
customer_id                0
transaction_date           0
online_order               0
order_status               0
brand                      0
product_line               0
product_class              0
product_size               0
list_price                 0
standard_cost              0
product_first_sold_date    0
dtype: int64

Percentage of Missing Values:
transaction_id             0.0
product_id                 0.0
customer_id                0.0
transaction_date           0.0
online_order               0.0
order_status               0.0
brand                      0.0
product_line               0.0
product_class              0.0
product_size               0.0
list_price                 0.0
standard_cost              0.0
product_first_sold_date    0.0
dtype: float64


**Converting datatypes**

In [156]:
# Convert 'list_price' column to a numeric data type
transaction_data['list_price'] = pd.to_numeric(transaction_data['list_price'])

# Convert 'standard_cost' column to a numeric data type
transaction_data['standard_cost'] = pd.to_numeric(transaction_data['standard_cost'])

# Convert 'transaction_date' column to a datetime data type
transaction_data['transaction_date'] = pd.to_datetime(transaction_data['transaction_date'])

In [157]:
# Print out the dataset information
transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19803 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           19803 non-null  int64         
 1   product_id               19803 non-null  int64         
 2   customer_id              19803 non-null  int64         
 3   transaction_date         19803 non-null  datetime64[ns]
 4   online_order             19803 non-null  float64       
 5   order_status             19803 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               19803 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [158]:
# Check for duplicate rows in the transaction_data dataset
duplicates = transaction_data[transaction_data.duplicated()]

# Count the number of duplicate rows
num_duplicates = len(duplicates)

# Display the duplicate rows and the total count
if num_duplicates > 0:
    print(f"Number of duplicate rows: {num_duplicates}")
    print("Duplicate rows:")
    print(duplicates)
else:
    print("No duplicate rows found in the transaction_data dataset.")

No duplicate rows found in the transaction_data dataset.


To ensure the consistency and quality of the categorical data in the dataset, we will create a list that includes all the categorical columns. Categorical columns are those that contain non-numeric data, such as labels, categories, or text values. This list will help us focus on examining and potentially cleaning these specific columns to maintain data integrity and accuracy during our analysis.

In [159]:
# Collect the categorical columns into a list
cat_col = transaction_data.select_dtypes(include='object').columns.tolist()
cat_col

['order_status', 'brand', 'product_line', 'product_class', 'product_size']

In [160]:
# Iterate through each categorical column in the dataset
for col in cat_col:
    # Print the column name
    print(f"Column: {col}")
    
    # Check for unique values in the column
    unique_values = transaction_data[col].unique()
    
    # Print the unique values
    print("Unique Values:")
    print(unique_values)
    
    # Check for duplicate values in the column
    is_duplicate = transaction_data[col].duplicated().any()
    
    # Print the duplicate status
    if is_duplicate:
        print("Duplicate values found in this column.")
    else:
        print("No duplicate values found in this column.")
    
    # Add a separator for clarity
    print("\n*******\n")

Column: order_status
Unique Values:
['Approved' 'Cancelled']
Duplicate values found in this column.

*******

Column: brand
Unique Values:
['Solex' 'Trek Bicycles' 'OHM Cycles' 'Norco Bicycles' 'Giant Bicycles'
 'WeareA2B']
Duplicate values found in this column.

*******

Column: product_line
Unique Values:
['Standard' 'Road' 'Mountain' 'Touring']
Duplicate values found in this column.

*******

Column: product_class
Unique Values:
['medium' 'low' 'high']
Duplicate values found in this column.

*******

Column: product_size
Unique Values:
['medium' 'large' 'small']
Duplicate values found in this column.

*******



It appears that the `product_first_sold_date` column may have an incorrect date format that needs to be corrected. To address this issue, we'll convert the data in this column to the correct datetime format.

In [161]:
# Convert 'product_first_sold_date' column to object (string) data type
transaction_data['product_first_sold_date'] = transaction_data['product_first_sold_date'].astype(str)

In [162]:
transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19803 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           19803 non-null  int64         
 1   product_id               19803 non-null  int64         
 2   customer_id              19803 non-null  int64         
 3   transaction_date         19803 non-null  datetime64[ns]
 4   online_order             19803 non-null  float64       
 5   order_status             19803 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               19803 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [163]:
# Convert 'product_first_sold_date' column to datetime with year, month, and date
transaction_data['product_first_sold_date'] = pd.to_datetime(transaction_data['product_first_sold_date'], unit='s').dt.strftime('%Y-%m-%d')

In [164]:

transaction_data.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost product_first_sold_date  
0       71.49          53.62              1970-01-01  
1     2091.47         388.92              1970-01-01  
2     1793.43         248.82              1970-01-01  
3     1198.46         381.10              1970-01-01  
4     1765.30         709.48              1970-01-01

In [165]:

# Create new variables based on 'transaction_date'
transaction_data['Transaction_year'] = transaction_data['transaction_date'].dt.year
transaction_data['Transaction_month'] = transaction_data['transaction_date'].dt.month_name()
transaction_data['Transaction_day'] = transaction_data['transaction_date'].dt.day
transaction_data['day_of_the_week'] = transaction_data['transaction_date'].dt.day_name()

# Convert the 'Transaction_year' and 'Transaction_day' columns to string data type
transaction_data['Transaction_year'] = transaction_data['Transaction_year'].astype(str)
transaction_data['Transaction_day'] = transaction_data['Transaction_day'].astype(str)

# Display the first few rows of the updated dataset
transaction_data.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost product_first_sold_date Transaction_year  \
0       71.49          53.62              1970-01-01             2017   
1     2091.47         388.92              1970-01-01             2017   
2     1793.43         248.82              1970-01-01             2017   
3     1198.46         381.10              1970-01-01             2017   
4     1765.30         709.48              1970-01-01             2017   

  Transaction_month Transaction_day day_of_the_week  
0          February              25        Saturday  
1               May              21          Sunday  
2           October              16          Monday  
3            August              31        Thursday  
4           October               1          Sunday

## Merging Datasets

Since the three datasets are interconnected, with the first one containing transaction information, the second one containing demographic data, and the third one containing customer addresses, we can establish relationships between them. We can use the 'customer_id' as a common key to link these datasets.

By performing an inner join, we combine these datasets based on the 'customer_id,' which acts as a primary key in the first dataset (transactions) and a foreign key in the second (demographics) and third (addresses) datasets. This process creates a unified dataset that contains transaction details along with corresponding demographic and address information for each customer. It allows us to analyze and draw insights from a comprehensive dataset that combines information from different sourc

In [173]:
# Joining the Transactions dataset with the Customer Demographics dataset using 'customer_id' as the common key.
# We perform an inner join to keep only the rows with matching customer IDs in both datasets.
transactions_demographics = transaction_data.merge(demographic_data, on='customer_id', how='inner')

# Display the first 5 rows of the merged dataset to inspect the result.
transactions_demographics.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1           11065           1         2950       2017-10-16           0.0   
2           18923          62         2950       2017-04-26           0.0   
3               2           3         3120       2017-05-21           1.0   
4            6862           4         3120       2017-10-05           0.0   

  order_status           brand product_line product_class product_size  ...  \
0     Approved           Solex     Standard        medium       medium  ...   
1     Approved  Giant Bicycles     Standard        medium       medium  ...   
2     Approved           Solex     Standard        medium       medium  ...   
3     Approved   Trek Bicycles     Standard        medium        large  ...   
4     Approved  Giant Bicycles     Standard          high       medium  ...   

   last_name  gender past_3_years_bike_related_purchases            job_title  \
0    Anthony    Male                                  19  Software Engineer I   
1    Anthony    Male                                  19  Software Engineer I   
2    Anthony    Male                                  19  Software Engineer I   
3  O'Donnell  Female                                  89  Clinical Specialist   
4  O'Donnell  Female                                  89  Clinical Specialist   

  job_industry_category wealth_segment deceased_indicator owns_car tenure Age  
0    Financial Services  Mass Customer                  N      Yes   10.0  68  
1    Financial Services  Mass Customer                  N      Yes   10.0  68  
2    Financial Services  Mass Customer                  N      Yes   10.0  68  
3                Health  Mass Customer                  N      Yes   10.0  44  
4                Health  Mass Customer                  N      Yes   10.0  44  

[5 rows x 28 columns]

In [174]:
# Merging the Transactions and Customer Demographics dataset with the Addresses dataset into a final_dataset dataset (Customers, Transactions, Addresses).
# We use 'customer_id' as the common key for the merge and perform an inner join to retain rows with matching customer IDs in all three datasets.
final_data = transactions_demographics.merge(customer_data, on='customer_id', how='inner')

# The resulting final dataset now contains information about customers, their transactions, and their addresses.

In [175]:
# print the information of the dataset
final_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19773 entries, 0 to 19772
Data columns (total 32 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   transaction_id                       19773 non-null  int64         
 1   product_id                           19773 non-null  int64         
 2   customer_id                          19773 non-null  int64         
 3   transaction_date                     19773 non-null  datetime64[ns]
 4   online_order                         19773 non-null  float64       
 5   order_status                         19773 non-null  object        
 6   brand                                19773 non-null  object        
 7   product_line                         19773 non-null  object        
 8   product_class                        19773 non-null  object        
 9   product_size                         19773 non-null  object        
 10  list_price

## Saving the final_data as csv

In [176]:
# Define the file path where you want to save the final CTA dataset.
output_file_path = 'final_cta_data.csv'

# Save the CTA dataset to a CSV file.
final_data.to_csv(output_file_path, index=False)

# Display a message indicating that the dataset has been saved.
print(f"The final CTA dataset has been saved to '{output_file_path}'.")

The final CTA dataset has been saved to 'final_cta_data.csv'.
